cpp/src/arrow/compute/kernels/ree_util_internal.cc

```diff
@@ -60,6 +60,7 @@ Result<std::shared_ptr<ArrayData>> PreallocateRunEndsArray(
 Result<std::shared_ptr<ArrayData>> PreallocateValuesArray(
     const std::shared_ptr<DataType>& value_type, bool has_validity_buffer, int64_t length,
     int64_t null_count, MemoryPool* pool, int64_t data_buffer_size) {
+  DCHECK(!has_validity_buffer || null_count != 0);
```

Comment:
It's difficult to understand for me... How about this style?

```suggestion
  DCHECK(!(has_validity_buffer && null_count == 0));
```

And how about adding the pre-condition in the doc as a comment here too?

Reply:
`!condition || consequence` is the more direct equivalent of `condition implies consequence`.

I'm going to add a comment saying `has_validity_buffer implies null_count != 0`.

https://en.wikipedia.org/wiki/Logical_equivalence#Logical_equivalences_involving_conditional_statements

Reply:
The more I think of it, I realize I can remove the `has_validity_buffer` parameter and infer it completely from the `null_count`.

Reply:
I'm removing `null_count` instead and having the callers change the `null_count`s from `kUnknownNullCount` to the correct value when they see fit. Seems like a more robust way to guarantee correctness.